core/src/main/java/com/opensymphony/xwork2/ognl/accessor/CompoundRootAccessor.java

```diff
@@ -233,24 +234,22 @@ public Object callMethod(Map context, Object target, String name, Object[] objec
 
             if ((argTypes == null) || !invalidMethods.containsKey(mc)) {
                 try {
-                    Object value = OgnlRuntime.callMethod((OgnlContext) context, o, name, objects);
-
-                    if (value != null) {
-                        return value;
-                    }
+                    return OgnlRuntime.callMethod((OgnlContext) context, o, name, objects);
                 } catch (OgnlException e) {
                     // try the next one
-                    Throwable reason = e.getReason();
+                    reason = e.getReason();
 
-                    if (!context.containsKey(OgnlValueStack.THROW_EXCEPTION_ON_FAILURE) && (mc != null) && (reason != null) && (reason.getClass() == NoSuchMethodException.class)) {
+                    if ((mc != null) && (reason != null) && (reason.getClass() == NoSuchMethodException.class)) {
                         invalidMethods.put(mc, Boolean.TRUE);
-                    } else if (reason != null) {
-                        throw new MethodFailedException(o, name, e.getReason());
                     }
                 }
             }
         }
 
+        if (context.containsKey(OgnlValueStack.THROW_EXCEPTION_ON_FAILURE)) {
+            throw new MethodFailedException(target, name, reason);
```

Comment:
Seems like previously this `MethodFailedException` is thrown when the method is found but failed on execution. Now we are mixing up both cases "no such method" and "method failed" by throwing this exception outside the loop. Is this the expected behavior?
Also, shall we consider break the loop at the point we threw `MethodFailedException`?

Reply:
No, previously it was mixed up both cases also so it is same as previous in this behavior. The only difference for it is: Now it is outside of the loop. Being inside of the loop is a bug. I discovered it in #357 - For example `OgnlValueStack.findValue(expr: "someProp", throwExceptionOnFailure: true)` always fail when top of stack doesn't have `someProp` property while it must check also other objects in stack and fail if none of them had this property.

> Also, shall we consider break the loop at the point we threw MethodFailedException?

I think no. AFAIK it must try on all of objects in stack.

Reply:
Looks ok to me, it makes sense to keep the mixed up behavior. 
Regarding the exception previously at line 248, doesn't that break the loop already? I don't think that would be an issue even `throwExceptionOnFailure` is set true. Because that is only thrown when the property is found, but an exception is thrown in application code (that's why `reason != null`).
Do you propose to search down the stack anyway even we've already "hit"?
The only use case I can think of: we have more than one "hit" on stack, while first one throws runtime exception and second one execute smoothly and return -- yes, in this case we don't want to break the loop.
I'm not super familiar with the coding style and design pattern of `ognl`, is that really what we want? 
Hiding exceptions doesn't look quite right to me...

Reply:
Now, it doesn't break the loop but previously it did, and I thought it was an issue when `throwExceptionOnFailure` is set true because of casual failed tests #357 - for example it accidentally uncovered that `XWorkConverter.getConversionErrorMessage` seems fails when user set `throwExceptionOnFailure` to true, I think.  right?

I just found all usages of `OgnlValueStack.THROW_EXCEPTION_ON_FAILURE` and found all of them outside loop except here. Thanks for heads up for `reason != null`, it makes sense! But now I'm surprised if `reason != null` is a good decision to distinguish between `property not found` and `user app exception` then why #357 tests fail?! (see prev para) In #357 everything is same as before except `throwExceptionOnFailure` which is hard-coded to true.